### PR TITLE
Add isPreferred to the CodeAction protocol.

### PR DESCRIPTION
**Pull request by Jason3S**
See [vscode isPreferred discussion](https://github.com/Microsoft/vscode/issues/62110#issuecomment-454182230)

### CODE REVIEW

**Comment by dbaeumer** [comment]
For this to be merged we need to update the spec as well (see https://github.com/Microsoft/language-server-protocol/blob/gh-pages/specification.md#L1). In addition I think this needs to be guarded by a client capability so the server can decided if it is worthwhile to compute the property (in this case inexpensive but this is how the spec works today).

**Comment by Jason3S** [comment]
I have updated the spec and I added `isPreferred` to the Code Action capabilities. I hope this is what you meant.

**Comment by Jason3S** [comment]
@dbaeumer Is there more I need to do?

File: types/src/main.ts [hunk 1]
```diff
@@ -2204,6 +2204,15 @@ export interface CodeAction {
 	 * executed and then the command.
 	 */
 	command?: Command;
+
+	/**
+     * Marks this as a preferred action. Preferred actions are used by the `auto fix` command and can be targeted
+     * by keybindings.
+     *
+     * A quick fix should be marked preferred if it properly addresses the underlying error.
+     * A refactoring should be marked preferred if it is the most reasonable choice of actions to take.
```
Comment (rcjsuen) [inline types/src/main.ts]: Is it just me or does this indenting look off?

Comment (Jason3S) [reply types/src/main.ts]: It is a tab vs spaces problem. Apparently this files uses tabs. I'll fix it.

**Comment by Jason3S** [comment]
Just waiting on: [Update specification.md Pull Request #736 · microsoft/language-server-protocol](https://github.com/microsoft/language-server-protocol/pull/736)

**Comment by dbaeumer** [comment]
Merged by hand with some minor modifications for consistency (e.g. `isPreferredSupport`).

**Pull request closed without merging**